nat-lab/bin/netcat.py

```diff
@@ -0,0 +1,209 @@
+                sys.exit(0)
+            print(
+                f"nc: connect to {hostname} port {port} ({self.sock_type}) failed: {e.strerror}",
+                file=sys.stderr,
+            )
+            sys.exit(1)
+
+    def _listen(self):
+        """Listen for incoming connections"""
+        # Set SO_REUSEPORT option
+        try:
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        except AttributeError:
+            # SO_REUSEPORT may not be available on all systems, fall back to SO_REUSEADDR
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+        port = self.args.port
+        bind_addr = "::" if self.ipv6 else "0.0.0.0"
+        self.sock.bind((bind_addr, port))
+
+        if self.udp:
+            self._vprint(f"Bound on {bind_addr} {port}")
+        else:
+            self.sock.listen(1)
+            self._vprint(f"Listening on {bind_addr} {port}")
+            conn, addr = self.sock.accept()
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.sock = conn
+
+    def _udptest(self):
+        """Test UDP connection"""
+        # Try sending data to the socket
+        for _ in range(UDP_SCAN_COUNT):
+            self.sock.send(b"X")
+            time.sleep(1)
+
+    def _register_socket(self):
+        """Register the socket to the selector for read events"""
+        self.selector.register(self.sock, selectors.EVENT_READ, self._read_from_socket)
+
+    def _register_stdin(self):
+        """Register stdin to the selector for read events, unless disabled"""
+        if not self.args.d:
+            self.selector.register(
+                sys.stdin, selectors.EVENT_READ, self._read_from_stdin
+            )
+
+    def _read_from_socket(self):
+        """Handle incoming data from the socket"""
+        data, addr = self.sock.recvfrom(RECV_SIZE)
+        if self.udp and self.listen and not self.client_addr:
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.client_addr = addr
+            self.sock.connect(addr)
+        if data:
+            sys.stdout.buffer.write(data)
+            sys.stdout.flush()
+        else:
+            # Connection closed
+            self.selector.unregister(self.sock)
+            self.should_close = True
+
+    def _read_from_stdin(self):
+        """Handle input from stdin and send it over the socket"""
+        data = sys.stdin.buffer.readline()
+        if data:
+            if self.listen and self.client_addr:
+                self.sock.sendto(data, self.client_addr)
+            else:
+                self.sock.send(data)
+        else:
+            # EOF on stdin
+            self.selector.unregister(sys.stdin)
+            self.should_close = True
+
+    def run(self):
+        try:
+            if self.args.l:
+                self._listen()
+            else:
+                self._connect()
+            if not self.args.z:
```

Comment:
-z seems to be unsupported. Maybe it is worth simply failing the startup with argparse error? IMHO it would make it clearer to see what is happening.

Reply:
Hmmm what do you mean, -z seems to be unsupported?
I tried to re-crate the behavior [form](https://sources.debian.org/src/netcat-openbsd/1.195-2/netcat.c/#L915) as closely as I could, and it seems to be the same from my tests and pcaps.

It tries to connect to the listener and then immediately closes and exits with 0 if it does, without reading anything from stdin, otherwise it "fails" with 1  from `Connection refused`

Reply:
Oopsie sorry, somehow I completely mixed up the meaning of `-z` flag. Yeah, you are right. I am resolving this comment.